### ISSUE TITLE
Repository Invalidation Failed With Limited Available Processors

### ISSUE THREAD

**Issue by yingnan** [oracle/opengrok#4202]
**Describe the bug**

When running OpenGrok on a container with cpu limit set to 1, indexer gets a 500 error from webapp:

```
Mar 03, 2023 3:16:13 AM org.opengrok.indexer.index.Indexer sendToConfigHost
SEVERE: Failed to send configuration to http://localhost:8080/ (is web application server running with opengrok deployed?)
java.io.IOException: InboundJaxrsResponse{context=ClientResponse{method=GET, uri=http://localhost:8080/api/v1/status/f2317aa3-577b-4c11-8d61-97d5403b728b, status=500, reason=Server Error}}
	at org.opengrok.indexer.configuration.RuntimeEnvironment.writeConfiguration(RuntimeEnvironment.java:1572)
	at org.opengrok.indexer.index.Indexer.sendToConfigHost(Indexer.java:1172)
	at org.opengrok.indexer.index.Indexer.main(Indexer.java:408)
```

Opengrok - 1.7.42 and 1.8.1
JDK - openjdk 17.0.6
OS - Rocky Linux 9.1
Jetty - 11.0.13
SCM - git

**To Reproduce**
* Limit container cpu to 1 and start OpenGrok.
* Run indexer with `-U` option and you should get the `sendToConfigHost` error mentioned above.
* Add webapp configuration `repositoryInvalidationParallelism=1` and restart OpenGrok.
* Repeat the indexer step and the `sendToConfigHost` error should not occur. 

**Expected behavior**
Indexing should be able to complete without requiring `repositoryInvalidationParallelism` option.

**Comment by yingnan** [comment]
My guess is that `HistoryGuru#invalidateRepositories` has trouble starting the thread pool when the following line can return 0.

https://github.com/oracle/opengrok/blob/5e6c01755d3bd778325fe6a3b2868adb7f02e9c1/opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/RuntimeEnvironment.java#L1180

**Comment by vladak** [comment]
What is the corresponding exception/error on the web app side ?

**Comment by yingnan** [comment]
> What is the corresponding exception/error on the web app side ?

No exception is thrown from the web app. I simply notice a big chunk of logs are missing from the web app after `PUT /api/v1/configuration` request.

Example logs when it failed:

```
FINE: API task {uuid=3df1552d-7705-4911-95a1-19aab1f95513,path=/api/v1/configuration,state=INITIAL,responseStatus=OK} started
Mar 09, 2023 7:18:45 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=6e550878-2a98-426c-86c2-612c9f148598, referenceCounter=0, store size=0} on thread qtp926370398-19
Mar 09, 2023 7:18:46 PM org.opengrok.indexer.history.HistoryGuru invalidateRepositories
FINE: invalidating 56 repositories
```

Comparing to logs when the request is successful:
```
FINE: API task {uuid=11757b08-dc25-4497-a706-9a6c4f8176df,path=/api/v1/configuration,state=INITIAL,responseStatus=OK} started
Mar 09, 2023 7:29:03 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=0a6b09b2-3798-4a8a-918f-3ed6b7262726, referenceCounter=0, store size=0} on thread qtp926370398-22
Mar 09, 2023 7:29:03 PM org.opengrok.indexer.history.HistoryGuru invalidateRepositories
FINE: invalidating 56 repositories
Mar 09, 2023 7:29:03 PM org.opengrok.web.AuthorizationFilter doFilter
FINER: Allowing request to /api/v1 in org.opengrok.web.AuthorizationFilter
Mar 09, 2023 7:29:03 PM org.opengrok.web.api.v1.filter.IncomingFilter filter
FINEST: allowing request to status/11757b08-dc25-4497-a706-9a6c4f8176df based on localhost IP address
Mar 09, 2023 7:29:03 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=becb0922-5acb-4c5e-bc03-c412c54785e9, referenceCounter=0, store size=0} on thread qtp926370398-21
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done invalidating repositories (56 valid, 56 working) (took 898 ms)
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.configuration.RuntimeEnvironment applyConfig
INFO: Configuration updated
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.configuration.RuntimeEnvironment maybeRefreshIndexSearchers
INFO: refreshing searcher managers
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done refreshing searcher managers (took 0 ms)
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.configuration.WatchDogService stop
INFO: Watchdog stopped
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.framework.PluginFramework reload
WARNING: Plugin directory not found or not readable: /home/opengrok/index/../plugins. All requests allowed.
Mar 09, 2023 7:29:04 PM org.opengrok.web.api.v1.filter.IncomingFilter onConfigurationChanged
FINER: refreshing token cache
Mar 09, 2023 7:29:04 PM org.opengrok.indexer.configuration.RuntimeEnvironment applyConfig
INFO: Done applying configuration
Mar 09, 2023 7:29:04 PM org.opengrok.web.api.v1.suggester.provider.service.impl.SuggesterServiceImpl refresh
FINE: Refreshing suggester for new configuration org.opengrok.indexer.configuration.SuggesterConfig@a779ca46
Mar 09, 2023 7:29:04 PM org.opengrok.web.AuthorizationFilter doFilter
FINER: Allowing request to /api/v1 in org.opengrok.web.AuthorizationFilter
Mar 09, 2023 7:29:04 PM org.opengrok.web.api.v1.filter.IncomingFilter filter
FINEST: allowing request to status/11757b08-dc25-4497-a706-9a6c4f8176df based on localhost IP address
Mar 09, 2023 7:29:04 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=49f2749d-7879-4482-9e3b-c4b9d2036741, referenceCounter=0, store size=0} on thread qtp926370398-18
Mar 09, 2023 7:29:06 PM org.opengrok.web.AuthorizationFilter doFilter
FINER: Allowing request to /api/v1 in org.opengrok.web.AuthorizationFilter
Mar 09, 2023 7:29:06 PM org.opengrok.web.api.v1.filter.IncomingFilter filter
FINEST: allowing request to status/11757b08-dc25-4497-a706-9a6c4f8176df based on localhost IP address
Mar 09, 2023 7:29:06 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=b17bc8d6-0943-4b7b-b2c3-4bf434f8444d, referenceCounter=0, store size=0} on thread qtp926370398-13
Mar 09, 2023 7:29:07 PM org.opengrok.web.AuthorizationFilter doFilter
FINER: Allowing request to /api/v1 in org.opengrok.web.AuthorizationFilter
Mar 09, 2023 7:29:07 PM org.opengrok.web.api.v1.filter.IncomingFilter filter
FINEST: allowing request to status/11757b08-dc25-4497-a706-9a6c4f8176df based on localhost IP address
Mar 09, 2023 7:29:07 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=b65a28e9-f543-49c7-888d-458c9185ee7f, referenceCounter=0, store size=0} on thread qtp926370398-19
Mar 09, 2023 7:29:08 PM org.opengrok.web.AuthorizationFilter doFilter
FINER: Allowing request to /api/v1 in org.opengrok.web.AuthorizationFilter
Mar 09, 2023 7:29:08 PM org.opengrok.web.api.v1.filter.IncomingFilter filter
FINEST: allowing request to status/11757b08-dc25-4497-a706-9a6c4f8176df based on localhost IP address
Mar 09, 2023 7:29:08 PM org.glassfish.jersey.internal.util.ExtendedLogger debugLog
FINEST: [DEBUG] Released scope instance Instance{id=60e8078c-ce5b-4a79-b8b6-dbe19385fc10, referenceCounter=0, store size=0} on thread qtp926370398-14
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: null
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration$$Lambda$983/0x0000000801274000@634e9dcc
Mar 09, 2023 7:29:09 PM java.io.ObjectInputStream filterCheck
FINER: ObjectInputFilter ALLOWED: class org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration, array length: -1, nRefs: 1, depth: 1, bytes: 115, ex: n/a
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: null
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration$$Lambda$983/0x0000000801274000@634e9dcc
Mar 09, 2023 7:29:09 PM java.io.ObjectInputStream filterCheck
FINER: ObjectInputFilter ALLOWED: class org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration, array length: -1, nRefs: 1, depth: 1, bytes: 115, ex: n/a
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: null
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration$$Lambda$983/0x0000000801274000@634e9dcc
Mar 09, 2023 7:29:09 PM java.io.ObjectInputStream filterCheck
FINER: ObjectInputFilter ALLOWED: class org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration, array length: -1, nRefs: 1, depth: 1, bytes: 115, ex: n/a
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: null
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration$$Lambda$983/0x0000000801274000@634e9dcc
Mar 09, 2023 7:29:09 PM java.io.ObjectInputStream filterCheck
FINER: ObjectInputFilter ALLOWED: class org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration, array length: -1, nRefs: 1, depth: 1, bytes: 115, ex: n/a
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: null
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration$$Lambda$983/0x0000000801274000@634e9dcc
Mar 09, 2023 7:29:09 PM java.io.ObjectInputStream filterCheck
FINER: ObjectInputFilter ALLOWED: class org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration, array length: -1, nRefs: 1, depth: 1, bytes: 115, ex: n/a
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: null
Mar 09, 2023 7:29:09 PM java.io.ObjectInputFilter$Config traceFilter
FINER: Builtin factory: null -> new: org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration$$Lambda$983/0x0000000801274000@634e9dcc
Mar 09, 2023 7:29:09 PM java.io.ObjectInputStream filterCheck
FINER: ObjectInputFilter ALLOWED: class org.opengrok.suggest.popular.impl.chronicle.ChronicleMapConfiguration, array length: -1, nRefs: 1, depth: 1, bytes: 115, ex: n/a
Mar 09, 2023 7:29:09 PM org.opengrok.suggest.Suggester lambda$getRebuildRunnable$2
FINE: Rebuild of SuggesterProjectData{indexDir=MMapDirectory@/home/opengrok/index/index/redacted lockFactory=org.apache.lucene.store.NativeFSLockFactory@7b08cd5, suggesterDir=/home/opengrok/index/suggester/redacted, allowMostPopular=true} finished, took PT8.463792016S
Mar 09, 2023 7:29:09 PM org.opengrok.suggest.Suggester shutdownAndAwaitTermination
INFO: Suggesters for [redacted] were successfully rebuilt (took 8 seconds)
Mar 09, 2023 7:29:09 PM org.opengrok.web.api.v1.suggester.provider.service.impl.SuggesterServiceImpl initSuggester
FINER: Suggester rebuild parallelism level: 1
Mar 09, 2023 7:29:09 PM org.opengrok.web.api.ApiTask lambda$getCallable$0
FINE: API task {uuid=11757b08-dc25-4497-a706-9a6c4f8176df,path=/api/v1/configuration,state=COMPLETED,responseStatus=OK} done
``` 

More specifically, I am looking for any of `INFO: Done invalidating repositories ...`, `INFO: Done applying configuration`, or `path=/api/v1/configuration,state=COMPLETED,responseStatus=OK`, but can't find any. That's why I think the issue is in `HistoryGuru#invalidateRepositories`.
